src/pyodide/python-entrypoint-helper.js

```diff
@@ -109,35 +119,66 @@ function pyimportMainModule(pyodide) {
   return pyodide.pyimport(mainModuleName);
 }
 
+let pyodidePromise;
+function getPyodide() {
+  if (pyodidePromise) {
+    return pyodidePromise;
+  }
+  pyodidePromise = loadPyodide(LOCKFILE, WORKERD_INDEX_URL);
+  return pyodidePromise;
+}
+
 let mainModulePromise;
-function getPyodide(ctx) {
-  if (mainModulePromise !== undefined) {
+function getMainModule() {
+  if (mainModulePromise) {
     return mainModulePromise;
   }
   mainModulePromise = (async function () {
-    // TODO: investigate whether it is possible to run part of loadPyodide in top level scope
-    // When we do it in top level scope we seem to get a broken file system.
-    const pyodide = await loadPyodide(ctx, LOCKFILE, WORKERD_INDEX_URL);
-    patchLoadPackage(pyodide);
-    const requirements = MetadataReader.getRequirements().map(
-      canonicalizePackageName,
-    );
-    const transitiveRequirements = new Set(
-      Array.from(await getTransitiveRequirements(pyodide, requirements)).map(
-        canonicalizePackageName,
-      ),
-    );
-    await setupPackages(pyodide, transitiveRequirements);
-    const mainModule = pyimportMainModule(pyodide);
-    return { mainModule };
+    const pyodide = await getPyodide();
+    await setupPackages(pyodide);
+    return pyimportMainModule(pyodide);
   })();
   return mainModulePromise;
 }
 
+if (IS_WORKERD) {
+  // If we're in workerd, we have to do setupPackages in the IoContext, so don't start it yet.
+  // TODO: fix this.
+  await getPyodide();
+} else {
+  // If we're not in workerd, setupPackages doesn't require IO so we can do it all here.
+  await getMainModule();
+}
+
+/**
+ * Have to reseed randomness in the IoContext of the first request since we gave a low quality seed
+ * when it was seeded at top level.
+ */
```

Comment:
It's probably a good idea to reseed the randomness for memory snapshots too, otherwise I think everyone gets the same sequence of random values.